tests/test_runtime.py

```diff
@@ -0,0 +1,667 @@
+        stride_bk: i32_t,
+        stride_bn: i32_t,
+        stride_cm: i32_t,
+        stride_cn: i32_t,
+    ):
+        pid = tl.program_id(axis="x")
+        num_pid_m = tl.cdiv(M, BLOCK_SIZE_M)
+        num_pid_n = tl.cdiv(N, BLOCK_SIZE_N)
+        num_pid_in_group = GROUP_SIZE_M * num_pid_n
+        group_id = pid // num_pid_in_group
+        first_pid_m = group_id * GROUP_SIZE_M
+        group_size_m = min(num_pid_m - first_pid_m, GROUP_SIZE_M)
+        pid_m = first_pid_m + (pid % group_size_m)
+        pid_n = (pid % num_pid_in_group) // group_size_m
+
+        # offs_am = (pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)) % M
+        # offs_bn = (pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)) % N
+        offs_am = pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+        offs_bn = pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+        offs_k = tl.arange(0, BLOCK_SIZE_K)
+        a_ptrs = a_ptr + (offs_am[:, None] * stride_am + offs_k[None, :] * stride_ak)
+        b_ptrs = b_ptr + (offs_k[:, None] * stride_bk + offs_bn[None, :] * stride_bn)
+
+        accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=float64)
+        acc = accumulator
+
+        for k, (acc, aptrs, bptrs) in range_(
+            0, tl.cdiv(K, BLOCK_SIZE_K), iter_args=[accumulator, a_ptrs, b_ptrs]
+        ):
+            mask = offs_k[None, :] < K - k * BLOCK_SIZE_K
+            a = tl.load(a_ptrs, mask=mask, other=0.0)
+            mask = offs_k[:, None] < K - k * BLOCK_SIZE_K
+            b = tl.load(b_ptrs, mask=mask, other=0.0)
+            # TODO(max): the problem here is the _update_frame_vars upstream
+            acc_next = acc + tl.dot(a, b)
+            aptrs_next = aptrs + BLOCK_SIZE_K * stride_ak
+            bptrs_next = bptrs + BLOCK_SIZE_K * stride_bk
+            yield_(acc_next, aptrs_next, bptrs_next)
+
+        c = acc
+
+        offs_cm = pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
+        offs_cn = pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
+        c_ptrs = c_ptr + stride_cm * offs_cm[:, None] + stride_cn * offs_cn[None, :]
+        c_mask = (offs_cm[:, None] < M) & (offs_cn[None, :] < N)
+        tl.store(c_ptrs, c, mask=c_mask)
+
+    matmul_kernel.emit()
+    module = backend.compile(
+        ctx.module,
+        kernel_name="matmul_kernel",
+        pipeline=Pipeline().add_pass("triton-to-linalg"),
+        generate_kernel_wrapper=False,
+        generate_return_consumer=False,
+    )
+
+    tensor_store = find_ops(
+        module.operation,
+        lambda op: op.operation.name == "memref.tensor_store",
+        single=True,
+    )
+    memref = to_memref(tensor_store.memref.type, tensor_store.tensor)
+    memref.owner.move_after(tensor_store)
+    c = copy(memref, tensor_store.memref)
+    c.move_after(memref.owner)
+    tensor_store.operation.erase()
+
+    module = backend.compile(
+        module,
+        kernel_name="matmul_kernel",
+        pipeline=Pipeline()
+        .bufferize()
```

Comment:
Hi @makslevental, does the `.bufferize()` here call the builtin one-shot bufferization?

Reply:
> Hi @makslevental, does the `.bufferize()` here call the builtin one-shot bufferization?

no https://github.com/makslevental/mlir-python-utils/blob/runtime/mlir_utils/runtime/passes.py#L132-L135